database/migrations/community/2024_12_16_000003_update_forumtopiccomment_table.php

```diff
@@ -0,0 +1,109 @@
+<?php
+
+use Illuminate\Database\Migrations\Migration;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Database\SQLiteConnection;
+use Illuminate\Support\Facades\Schema;
+
+return new class() extends Migration {
+    public function up(): void
+    {
+        if (DB::connection() instanceof SQLiteConnection) {
+            // do nothing
+        } else {
+            Schema::table('ForumTopicComment', function (Blueprint $table) {
+                // Check for both possible foreign key names since they could differ
+                // between production and local environments after migrations/rollbacks.
+                $foreignKeys = DB::select(<<<SQL
+                    SELECT CONSTRAINT_NAME
+                    FROM information_schema.TABLE_CONSTRAINTS
+                    WHERE TABLE_NAME = "ForumTopicComment"
+                    AND CONSTRAINT_TYPE = "FOREIGN KEY"
+                SQL);
+
+                foreach ($foreignKeys as $fk) {
+                    $table->dropForeign($fk->CONSTRAINT_NAME);
+                }
```

Comment:
This was not happy. Perhaps I'm still in a bad state from the previous attempts?
```
SQLSTATE[42000]: Syntax error or access violation: 1091 Can't DROP FOREIGN KEY `forumtopiccomment_author_id_foreign`; check that it exists (Connection: mysql, SQL: alter table `ForumTopicComment` drop foreign key `forumtopiccomment_author_id_foreign`)
```
```
  PRIMARY KEY (`ID`),
  KEY `forum_topic_comments_forum_topic_id_index` (`ForumTopicID`),
  KEY `forum_topic_comments_created_at_index` (`DateCreated`),
  KEY `forum_topic_comments_author_id_created_at_index` (`author_id`,`DateCreated`) USING BTREE,
  KEY `forumtopiccomment_forumtopicid_authorised_datecreated_index` (`ForumTopicID`,`Authorised`,`DateCreated`)
```

Reply:
It could be, but I'm also seeing something really weird on stage:
```sql
SELECT 
    CONSTRAINT_NAME,
    COLUMN_NAME,
    REFERENCED_TABLE_NAME,
    REFERENCED_COLUMN_NAME
FROM information_schema.KEY_COLUMN_USAGE
WHERE TABLE_NAME = 'ForumTopicComment'
AND REFERENCED_TABLE_NAME IS NOT NULL;
```
![Screenshot 2024-12-21 at 4 49 32 PM](https://github.com/user-attachments/assets/317fc1a4-49ae-4f4c-8aea-d4d6f64e3395)

It seems this constraint exists three times?? I am very confused by this query result, but I suspect that because the `SELECT` in the migration on your local is returning duplicative results, the loop is trying to drop the same constraint multiple times.

I'm not sure why I didn't encounter this locally. I've made some tweaks to the migration though to make it a bit more resilient against this sort of thing:

```php
$foreignKeys = DB::select(<<<SQL
    SELECT DISTINCT CONSTRAINT_NAME -- !! <------ add DISTINCT here to filter out duplicates
    FROM information_schema.TABLE_CONSTRAINTS
    WHERE TABLE_NAME = "ForumTopicComment"
    AND CONSTRAINT_TYPE = "FOREIGN KEY"
SQL);

foreach ($foreignKeys as $fk) {
    try {
        $table->dropForeign($fk->CONSTRAINT_NAME);
    } catch (Exception $e) { // !! <----------------------------- also added try/catch
        // If dropping the constraint fails, just continue to the next one.
        continue;
    }
}
```

Reply:
I'm still seeing this error. I'm going to assume my DB is broken and reset it.
```
mysql> SELECT DISTINCT CONSTRAINT_NAME
    ->                     FROM information_schema.TABLE_CONSTRAINTS
    ->                     WHERE TABLE_NAME = "ForumTopicComment"
    ->                     AND CONSTRAINT_TYPE = "FOREIGN KEY";
+-------------------------------------+
| CONSTRAINT_NAME                     |
+-------------------------------------+
| forumtopiccomment_author_id_foreign |
+-------------------------------------+
1 row in set (0.00 sec)
mysql> show create table ForumTopicComment;
+-------------------+-------------------------------
| Table             | Create Table                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   |
+-------------------+-------------------------------
| ForumTopicComment | CREATE TABLE `ForumTopicComment` (
  `ID` bigint(20) unsigned NOT NULL AUTO_INCREMENT,
  `ForumTopicID` bigint(20) unsigned DEFAULT NULL,
  `Payload` text NOT NULL,
  `author_id` bigint(20) unsigned DEFAULT NULL,
  `authorized_at` timestamp NULL DEFAULT NULL,
  `DateCreated` timestamp NULL DEFAULT NULL,
  `DateModified` timestamp NULL DEFAULT current_timestamp() ON UPDATE current_timestamp(),
  `Authorised` tinyint(3) unsigned DEFAULT NULL,
  `deleted_at` timestamp NULL DEFAULT NULL,
  PRIMARY KEY (`ID`),
  KEY `forum_topic_comments_forum_topic_id_index` (`ForumTopicID`),
  KEY `forum_topic_comments_created_at_index` (`DateCreated`),
  KEY `forum_topic_comments_author_id_created_at_index` (`author_id`,`DateCreated`) USING BTREE,
  KEY `forumtopiccomment_forumtopicid_authorised_datecreated_index` (`ForumTopicID`,`Authorised`,`DateCreated`)
) ENGINE=InnoDB AUTO_INCREMENT=246383 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_uca1400_ai_ci |
+-------------------+-------------------------------
```

Reply:
Tomorrow I'll also retry this from a fresh prod db image.

Reply:
I've confirmed that from a fresh prod DB snapshot, the four migrations complete successfully.

Reply:
I _think_ this is working with a fresh dump. It failed a bit lower due to not running the before deployment SQL, and now this step is partially deployed and I can't proceed or rollback. Going to reimport the dump again.

Reply:
Before this PR ever gets merged, we should probably run the before deployment SQL on both stage and prod as a safety measure.

Reply:
Worked fine after running the before-SQL.